File: ruby/lib/services/account_service.rb
# frozen_string_literal: true

require 'services/service_base'

module Trinsic
  # Account Service wrapper
  class AccountService < ServiceBase
    def initialize(service_options = nil)
      super(service_options)
      if @service_options.server_use_tls
        channel_creds = GRPC::Core::ChannelCredentials.new
        @client = Account::Account::Stub.new(url_string, channel_creds)
      else
        @client = Account::Account::Stub.new(url_string, :this_channel_is_insecure)
      end
    end

    def sign_in(request = nil)
      request ||= Account::SignInRequest.new
      request.details = request.details || Account::AccountDetails.new
      request.ecosystem_id = request.ecosystem_id || 'default'
      auth_token = @client.sign_in(request).profile
      encoded_profile = Base64.urlsafe_encode64(Account::AccountProfile.encode(auth_token))
      @token_provider.save(encoded_profile) unless auth_token.protection.enabled
      encoded_profile
    end

    def unprotect(profile, security_code)
      cloned = profile.clone
      request = Okapi::Security::V1::UnBlindOberonTokenRequest.new(token: cloned.auth_token)
      request.blinding += [security_code]
      result = Okapi::Oberon.unblind_token request
      cloned.auth_token = result.token
      cloned.protection = Account::TokenProtection.new(enabled: false, method: Account::ConfirmationMethod.None)
      cloned
    end

    def protect(profile, security_code)
      cloned = profile.clone
      request = Okapi::Security::V1::BlindOberonTokenRequest.new(token: cloned.auth_token)
      request.blinding += [security_code]
      result = Okapi::Oberon.blind_token request
      cloned.auth_token = result.token
      cloned.protection = Account::TokenProtection.new(enabled: true, method: Account::ConfirmationMethod.Other)
      cloned
    end

    def login(request = nil)
      request ||= Account::LoginRequest.new
      response = @client.login(request)
      unless response.profile.nil?
        encoded_profile = Base64.urlsafe_encode64(Account::AccountProfile.encode(response.profile))
        @token_provider.save(encoded_profile)
      end
      response
    end

    def login_confirm(challenge, auth_code)
      hashed = Okapi::Hashing.blake3_hash(Okapi::Hashing::V1::Blake3HashRequest.new(data: auth_code))
      request = Account::LoginConfirmRequest.new(challenge: challenge, confirmation_code_hashed: hashed.digest)
      response = @client.login_confirm(request)
      return nil if response.profile.nil?

      profile = response.profile
      profile = unprotect(profile, auth_code) if response.profile.protection.enabled
      encoded_profile = Base64.urlsafe_encode64(Account::AccountProfile.encode(profile))
      @token_provider.save(encoded_profile)
      encoded_profile
    end

    def login_anonymous(ecosystem_id = nil)
      ecosystem_id ||= 'default'

      request = Account::LoginRequest.new(ecosystem_id: ecosystem_id)
      response = login(request)

      raise Error('nil profile returned') if response.profile.nil?
      raise Error('protected profile returned') if response.profile.protection.enabled

      encoded_profile = Base64.urlsafe_encode64(Account::AccountProfile.encode(response.profile))
      @token_provider.save(encoded_profile)
      encoded_profile
    end

    def get_info
      info
    end

    # BEGIN Code generated by protoc-gen-trinsic. DO NOT EDIT.
    # target: /home/runner/work/sdk/sdk/ruby/lib/services/account_service.rb
    nil

    #  Get account information
    # @param [AccountInfoRequest] request
    # @return [Services::Account::V1::AccountInfoResponse]
    def info
      request = Services::Account::V1::AccountInfoRequest.new
      @client.info(request, metadata: metadata(request))
    end

    #  List all connected devices
    # @param [ListDevicesRequest] request
    # @return [Services::Account::V1::ListDevicesResponse]
    def list_devices(request)
      @client.list_devices(request, metadata: metadata(request))
    end

    #  Revoke device access to the account's cloud wallet
    # @param [RevokeDeviceRequest] request
    # @return [Services::Account::V1::RevokeDeviceResponse]
    def revoke_device(request)
      @client.revoke_device(request, metadata: metadata(request))
    end

    #  Authorize Ecosystem to receive webhook events
    # @param [AuthorizeWebhookRequest] request
    # @return [Services::Account::V1::AuthorizeWebhookResponse]
    def authorize_webhook(request)
      @client.authorize_webhook(request, metadata: metadata(request))
    end

    # END Code generated by protoc-gen-trinsic. DO NOT EDIT.
  end
end

File: go/services/account_service.go
package services

import (
	"context"
	"encoding/base64"
	"errors"

	"github.com/trinsic-id/okapi/go/okapi/hashing/v1/hashing"
	"github.com/trinsic-id/okapi/go/okapi/security/v1/security"

	"github.com/trinsic-id/okapi/go/okapi"
	"github.com/trinsic-id/sdk/go/proto/services/account/v1/account"

	"google.golang.org/protobuf/proto"
)

// NewAccountService returns an account service with the base service configured
// using the provided options
func NewAccountService(options *Options) (AccountService, error) {
	base, err := NewServiceBase(options)
	if err != nil {
		return nil, err
	}
	service := &accountBase{
		Service: base,
		client:  account.NewAccountClient(base.GetChannel()),
	}

	return service, nil
}

// AccountService wraps all the functions for interacting with accounts
type AccountService interface {
	Service

	// SignIn (DEPRECATED, USE Login) attempts to log in and returns an authentication token
	SignIn(userContext context.Context, request *account.SignInRequest) (string, account.ConfirmationMethod, error)

	// Unprotect takes an authtoken that has been protected using a pin code
	// and returns an unlocked token
	Unprotect(authtoken, securityCode string) (string, error)

	// Protect will apply the given security code blind to the provided token
	Protect(authtoken, securityCode string) (string, error)

	// Login performs the first stage of login, which must be finalized with LoginConfirm
	Login(userContext context.Context, request *account.LoginRequest) (*account.LoginResponse, error)

	// LoginConfirm finalizes login, using challenge received from Login(), and authCode sent to user email
	LoginConfirm(userContext context.Context, challenge []byte, authCode string) (string, error)

	// LoginAnonymous creates an anonymous account in the given ecosystem and returns an auth token
	LoginAnonymous(userContext context.Context, ecosystemId ...string) (string, error)

	// GetInfo returns details about the wallet associated with the account token
	GetInfo(userContext context.Context) (*account.AccountInfoResponse, error)

	// BEGIN Interface Code generated by protoc-gen-trinsic. DO NOT EDIT.
	// target: /home/runner/work/sdk/sdk/go/services/account_service.go

	// Info  Get account information
	Info(userContext context.Context) (*account.AccountInfoResponse, error)
	// AuthorizeWebhook  Authorize Ecosystem to receive webhook events
	AuthorizeWebhook(userContext context.Context, request *account.AuthorizeWebhookRequest) (*account.AuthorizeWebhookResponse, error)

	// END Interface Code generated by protoc-gen-trinsic. DO NOT EDIT.
}

type accountBase struct {
	Service
	client account.AccountClient
}

// SignIn to a given account
func (a *accountBase) SignIn(userContext context.Context, request *account.SignInRequest) (string, account.ConfirmationMethod, error) {
	if request == nil {
		request = &account.SignInRequest{}
	}
	if request.Details == nil {
		request.Details = &account.AccountDetails{}
	}

	resp, err := a.client.SignIn(userContext, request)
	if err != nil {
		return "", account.ConfirmationMethod_None, err
	}

	authToken, err := ProfileToToken(resp.Profile)
	if err != nil {
		return "", account.ConfirmationMethod_None, err
	}

	a.SetAuthToken(authToken)
	a.GetTokenProvider().SaveDefault(authToken)

	return authToken, resp.ConfirmationMethod, nil
}

// Unprotect an authtoken using the given security code
//
// This method can be called multiple times on an individual token
// to "unwrap" the blindings that have been applied
func (a *accountBase) Unprotect(authtoken, securityCode string) (string, error) {
	profile, err := ProfileFromToken(authtoken)
	if err != nil {
		return "", err
	}

	request := &security.UnBlindOberonTokenRequest{
		Token:    profile.AuthToken,
		Blinding: append([][]byte{}, []byte(securityCode)),
	}

	response, err := okapi.Oberon().UnBlindToken(request)
	if err != nil {
		return "", err
	}

	profile.AuthToken = response.Token
	profile.Protection = &account.TokenProtection{
		Enabled: false,
		Method:  account.ConfirmationMethod_None,
	}

	return ProfileToToken(profile)
}

// Protect an authtoken with the given security code. Must be unprotected before use
//
// This method can be called as many times as you want, but each code must be "unwrapped"
// by calling Unprotect in the reverse order before use
func (a *accountBase) Protect(authtoken, securityCode string) (string, error) {
	profile, err := ProfileFromToken(authtoken)
	if err != nil {
		return "", err
	}

	request := &security.BlindOberonTokenRequest{
		Token:    profile.AuthToken,
		Blinding: append([][]byte{}, []byte(securityCode)),
	}

	response, err := okapi.Oberon().BlindToken(request)
	if err != nil {
		return "", err
	}

	profile.AuthToken = response.Token
	profile.Protection = &account.TokenProtection{
		Enabled: true,
		Method:  account.ConfirmationMethod_Other,
	}

	return ProfileToToken(profile)
}

func (a *accountBase) Login(userContext context.Context, request *account.LoginRequest) (*account.LoginResponse, error) {
	response, err := a.client.Login(userContext, request)
	if err != nil {
		return nil, err
	}

	return response, nil
}

func (a *accountBase) LoginConfirm(userContext context.Context, challenge []byte, authCode string) (string, error) {
	// Hash the authcode
	codeHash, err := okapi.Hashing().Blake3Hash(&hashing.Blake3HashRequest{Data: []byte(authCode)})
	if err != nil {
		return "", err
	}

	// Generate request
	request := &account.LoginConfirmRequest{
		Challenge:              challenge,
		ConfirmationCodeHashed: codeHash.Digest,
	}

	// Send request
	response, err := a.client.LoginConfirm(userContext, request)
	if err != nil {
		return "", err
	}

	// If profile isn't protected, something is wrong.
	if !response.Profile.Protection.Enabled {
		return "", errors.New("account information was not protected")
	}

	// Tokenize profile
	authToken, err := ProfileToToken(response.Profile)
	if err != nil {
		return "", err
	}

	// Unprotect profile
	authToken, err = a.Unprotect(authToken, authCode)
	if err != nil {
		return "", err
	}

	a.SetAuthToken(authToken)
	a.GetTokenProvider().SaveDefault(authToken)

	return authToken, nil
}

func (a *accountBase) LoginAnonymous(userContext context.Context, ecosystemId ...string) (string, error) {
	// `ecosystemId` is a `...string` to emulate default-value argument behavior, which Go doesn't support.
	// We'll just use the first index of `ecosystemId`; if it's an empty array, we use the default value
	ecoId := ""
	if len(ecosystemId) > 0 {
		ecoId = ecosystemId[0]
	}

	// Create request
	request := &account.LoginRequest{
		Email:       "",
		EcosystemId: ecoId,
	}

	// Attempt login
	response, err := a.Login(userContext, request)
	if err != nil {
		return "", err
	}

	// Interrogate results
	if response.GetProfile() == nil {
		return "", errors.New("no profile returned from Login()")
	}
	if response.GetProfile().Protection.Enabled {
		return "", errors.New("protected profile returned from Login()")
	}

	// Tokenize and return profile
	authToken, err := ProfileToToken(response.GetProfile())
	if err != nil {
		return "", err
	}

	a.SetAuthToken(authToken)
	a.GetTokenProvider().SaveDefault(authToken)

	return authToken, nil
}

// GetInfo associated with a given wallet
func (a *accountBase) GetInfo(userContext context.Context) (*account.AccountInfoResponse, error) {
	request := &account.AccountInfoRequest{}
	md, err := a.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}
	response, err := a.client.Info(md, request)
	if err != nil {
		return nil, err
	}
	return response, nil
}

// ProfileToToken takes the proile and returns an encoded auth token
func ProfileToToken(profile *account.AccountProfile) (string, error) {
	pbytes, err := proto.Marshal(profile)
	if err != nil {
		return "", err
	}

	return base64.RawURLEncoding.EncodeToString(pbytes), nil
}

// ProfileFromToken takes an encoded auth token and returns the account profile
func ProfileFromToken(token string) (*account.AccountProfile, error) {
	tb, err := base64.RawURLEncoding.DecodeString(token)
	if err != nil {
		return nil, err
	}

	profile := &account.AccountProfile{}

	err = proto.Unmarshal(tb, profile)

	return profile, err
}

// BEGIN Implementation Code generated by protoc-gen-trinsic. DO NOT EDIT.
// target: /home/runner/work/sdk/sdk/go/services/account_service.go

// Info  Get account information
func (a *accountBase) Info(userContext context.Context) (*account.AccountInfoResponse, error) {
	request := &account.AccountInfoRequest{}
	md, err := a.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}
	response, err := a.client.Info(md, request)
	if err != nil {
		return nil, err
	}
	return response, nil
}

// AuthorizeWebhook  Authorize Ecosystem to receive webhook events
func (a *accountBase) AuthorizeWebhook(userContext context.Context, request *account.AuthorizeWebhookRequest) (*account.AuthorizeWebhookResponse, error) {
	md, err := a.GetMetadataContext(userContext, request)
	if err != nil {
		return nil, err
	}
	response, err := a.client.AuthorizeWebhook(md, request)
	if err != nil {
		return nil, err
	}
	return response, nil
}

// END Implementation Code generated by protoc-gen-trinsic. DO NOT EDIT.
